pkg/finality-grandpa/voting_round.go

```diff
@@ -0,0 +1,842 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"time"
+
+	"golang.org/x/exp/constraints"
+)
+
+type stateStart[T any] [2]T
+
+type stateProposed[T any] [2]T
+
+type statePrevoting[T, W any] struct {
+	T T
+	W W
+}
```

Comment:
When is T and W?

Reply:
In practice `T` is for timer, and `W` for `hashBestChain`.  Not sure why they named it `W` in the rust impl ([link](https://github.com/paritytech/finality-grandpa/blob/19ab8d7be65b3a19953ae070122f422c0e00619b/src/voter/voting_round.rs#L43)).

Reply:
My feeling is that these names are unintuitive enough that we should probably change them, even if it differs from the reference impl

Reply:
I think `T` for `Timer` makes sense but I don't think that's why they picked it.  `T` is usually used as the default first generic param.  I ended up changing the second param to `U`, since it's the next lexicographic latter, to signal as some other generic param.